bin/jetstream.php

```diff
@@ -0,0 +1,855 @@
+	$pid = intval(file_get_contents($pidfile));
+}
+
+if (empty($pid) && in_array($mode, ['stop', 'status'])) {
+	die("Pidfile wasn't found. Is jetstream running?\n");
+}
+
+if ($mode == 'status') {
+	if (posix_kill($pid, 0)) {
+		die("Jetstream process $pid is running.\n");
+	}
+
+	unlink($pidfile);
+
+	die("Jetstream process $pid isn't running.\n");
+}
+
+if ($mode == 'stop') {
+	posix_kill($pid, SIGTERM);
+
+	unlink($pidfile);
+
+	Logger::notice('Jetstream process was killed', ['pid' => $pid]);
+
+	die("Jetstream process $pid was killed.\n");
+}
+
+if (!empty($pid) && posix_kill($pid, 0)) {
+	die("Jetstream process $pid is already running.\n");
+}
+
+Logger::notice('Starting jetstream daemon.', ['pid' => $pid]);
+
+if (!$foreground) {
+	echo "Starting jetstream daemon.\n";
+
+	DBA::disconnect();
+
+	// Fork a daemon process
+	$pid = pcntl_fork();
+	if ($pid == -1) {
+		echo "Daemon couldn't be forked.\n";
+		Logger::warning('Could not fork daemon');
+		exit(1);
+	} elseif ($pid) {
+		// The parent process continues here
+		if (!file_put_contents($pidfile, $pid)) {
+			echo "Pid file wasn't written.\n";
+			Logger::warning('Could not store pid file');
+			posix_kill($pid, SIGTERM);
+			exit(1);
+		}
+		echo 'Child process started with pid ' . $pid . ".\n";
+		Logger::notice('Child process started', ['pid' => $pid]);
+		exit(0);
+	}
+
+	// We now are in the child process
+	register_shutdown_function('shutdown');
+
+	// Make the child the main process, detach it from the terminal
+	if (posix_setsid() < 0) {
+		return;
+	}
+
+	// Closing all existing connections with the outside
+	fclose(STDIN);
+
+	// And now connect the database again
+	DBA::connect();
+}
+
+// Just to be sure that this script really runs endlessly
+set_time_limit(0);
+
+// Now running as a daemon.
+require_once("addon/bluesky/bluesky.php");
+
+$jetstream = new Jetstream();
+$jetstream->listen();
+
+function shutdown()
+{
+	posix_kill(posix_getpid(), SIGTERM);
+	posix_kill(posix_getpid(), SIGHUP);
+}
+
+/**
+ * Class to handle the Bluesky Jetstream firehose
+ *
+ * Existing collections:
+ * app.bsky.feed.like, app.bsky.graph.follow, app.bsky.feed.repost, app.bsky.feed.post, app.bsky.graph.block,
+ * app.bsky.actor.profile, app.bsky.graph.listitem, app.bsky.graph.list, app.bsky.graph.listblock, app.bsky.feed.generator,
+ * app.bsky.feed.threadgate, app.bsky.graph.starterpack, app.bsky.feed.postgate, chat.bsky.actor.declaration,
+ * app.bsky.actor.domain, industries.geesawra.webpages
+ *
+ * Available servers:
+ * jetstream1.us-east.bsky.network, jetstream2.us-east.bsky.network, jetstream1.us-west.bsky.network, jetstream2.us-west.bsky.network
+ *
+ * @see https://github.com/bluesky-social/jetstream
+ * @todo Support more collections, support full firehose
+ */
+class Jetstream
+{
+	private $uids = [];
+	private $self = [];
+
+	public function listen()
+	{
+		$timeout       = 300;
+		$timeout_limit = 10;
+		$timestamp     = DI::keyValue()->get('jetstream_timestamp') ?? 0;
```

Comment:
> Maybe it's possible to test this class with given messages?

If we move the dependencies into the constructor it will be possible to create unit tests with mocks.

Reply:
`DI` calls are now no more in the class.